### ISSUE TITLE
[css-grid-1] Track Sizing Algorithm question

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#2873]
Issue copied from [original email](https://lists.w3.org/Archives/Public/www-style/2018Jul/0001.html) from **Jason.W He** (hekylinwork@gmail.com):

-------------

 
In case of following grid.

--- 65% -- | -- 35% -- |
min-content
--- 65% -- | -- 35% -- |
auto
--- 65% -- | -- 35% -- |

Two column: 65%, 35%
Two row: min-content, auto

template-areas:
"head side"
"body side"

For row #1's track size, which it use `min-content` to define height of that row. When calculate height, why is row #1 been calculated using items with span of 2?

From spec, it consider items with span of 1 first, then consider at span of 2 that do not span a track with a flexible sizing function. Which the second row `auto` get caught.

When change second row's auto to 1fr (flexible sizing function), it now has correct behaviour.

I have sample to reproduce this under
https://codepen.io/Nness/pen/MrYMzQ?editors=1100#

When increment items inside area `side`. The area `head`'s height now resize according to `side`.

Is this logical and correct behaviour to increase size of single span item by multiple span item?

It make sense to increase total size of multi-span grid track by item span across multiple track. How is that logical to change size of single-span grid track by item span across multiple track?

All the browser I have tested have exactly same result, does that means they all according to spec and have correct behaviour?

If use flexible sizing function is the only way to get this working and it is according to spec, then in which situation we want to use auto to have multiple span item change single span track's height behaviour?

**Comment by svillar** [comment]
> From spec, it consider items with span of 1 first, then consider at span of 2 that do not span a track with a flexible sizing function. Which the second row auto get caught.

This is not correct. The second row is auto sized. Flexible sized tracks are those using fr. See https://www.w3.org/TR/css-grid-1/#layout-algorithm

> When change second row's auto to 1fr (flexible sizing function), it now has correct behaviour.

Not sure what you mean with "correct" but I guess I've explained it in the previous comment.

**Comment by Nness** [comment]
This is correct behaviour. Which second row is 1fr
![2018-07-05_08h04_52](https://user-images.githubusercontent.com/1712143/42292911-284a76c8-802a-11e8-8747-a720df6239f4.png)

This is incorrect behaviour. Second row is auto
![2018-07-05_08h05_07](https://user-images.githubusercontent.com/1712143/42292913-29d3947a-802a-11e8-9ae1-b7959eaf844d.png)

The question is, is second one acceptable result? If it is according to spec, then what kind of scenario it trying to cover? Why do we want content inside item that span multiple tracks to increase height of single track item?

This is not related to total size, I mean individual size.
You can check on sample: https://codepen.io/Nness/pen/MrYMzQ?editors=1100#

Then add more `<div>push</div>` into `<div class="item-side">`

**Comment by fantasai** [comment]
I think it makes sense to fix this--to make min-content fit tighter than auto/max-content when possible.

**Comment by tabatkins** [comment]
Flagging for f2f. We think the *author expectation* of this case is fairly clear and the spec would be reasonably straightforward to fix, but at this point the compat impact is very worrying and requires analysis.

**Comment by Loirooriol** [comment]
The problem seems to be in this line: https://github.com/w3c/csswg-drafts/blob/0dcb036011e9ff870bc3f5de03d53210749ca0b6/css-grid-1/Overview.bs#L4425
It treats all intrinsic max track sizing functions equally instead of prioritizing `min-content` vs `auto/max-content`.

**Comment by tabatkins** [comment]
Okay, explaining the current behavior took some doing. [Here's a testcase showing it more simply:](http://software.hixie.ch/utilities/js/live-dom-viewer/saved/11888)

```html
<!doctype html>
<div class="container">
  <div class="item-head">
    <div class=block style="height: 20px;"></div>
  </div>
  <div class="item-body">
    <div class=block style="height: 80px;"></div>
  </div>
  <div class="item-side">
    <div class=block style="height: 150px"></div>
  </div>
</div>
<style>
.container {
  background-color: #fefefe;
  display: grid;
  grid-template-columns: 150px auto;
  grid-template-rows: min-content auto;
  grid-template-areas: 
    "head side"
    "body side";
}
.item-head {
  background-color: #d4edda;
  grid-area: head;
}
.item-body {
  background-color: #fff3cd;
  grid-area: body;
}
.item-side {
  background-color: #cce5ff;
  grid-area: side;
  overflow:hidden;
}
.block {
  width: 100px;
  background: #0002;
}
</style>
```

As you can see, the blue spanning item gives *all* of its content's height to the "min-content" track, and none to the "auto" track. This is the opposite of what the OP asks for, and is confusing to boot!

This happens because the blue item is scrollable (via `overflow:hidden`). This means it doesn't get an automatic minimum size. In [section 11.5, step 3.1](https://drafts.csswg.org/css-grid-1/#track-size-intrinsic-min), we distribute zero space, as that's the element's "minimum size". Then in step 3.2 we distribute min-content height, which is 150px, to the min-content track only - 150px to distribute, minus 100px of space already present in the track base sizes, means the "min-content" track grows by 50px. Then we're done, all items are considered and all space is distributed.

If the blue item is *not* scrollable, it has an auto min size of its min-content height, and so step 3.1 distributes that space to both tracks equally. We might still want to prioritize the auto track over the min-content track in this case. But it's not the problem OP is running into.

To address OP's problem, we need to address the fact that the auto track, which has absorbed the spanning item's minimum contribution (0px), isn't absorbing any of the spanning item's min-content contribution (150px). But it's not clear how to do that in the algorithm without breaking other expectations.

**Comment by fantasai** [comment]
Note: The desired behavior is likely more closely represented by this:
```
  grid-template-rows: minmax(auto, min-content) auto;
```
instead of
```
  grid-template-rows: min-content auto;
```

Still leaves open the idea of prioritizing auto/max-content tracks over min-content, though. Or other possible adjustments to track sizing.

**Comment by tabatkins** [comment]
<details>
 <summary>Notes for future Tab and fantasai</summary>

<p>In the spirit of fr tracks, let spanned auto maximum tracks absorb space over intrinsic minimums?

<p>Possible Idea:
<ul>
    <li>address all items not spanning fr or auto
    <li>address all items spanning auto
    <li>address all items spanning fr
</ul>
<p>Other Possible idea
<ul>
    <li>in distribute across spanning tracks, split "distribute up to limits" into two phases, preferentially distributing into auto tracks
    <li>maybe only do that for step 3.2 specifically, under some restrictive conditions?
</ul>
</details>

**Comment by Loirooriol** [comment]
I also analyzed the testcase with more detail after the meeting, but forgot to post my thoughts.

Basically, some confusing behavior of grid is that `auto` has completely opposite meanings when used as a min track sizing function vs a max track sizing function.

 - As a min track sizing function, it uses either the minimum contribution or the limited min-content contribution, which are both smaller than or equal to the min-content contribution. So the order is like `auto` <= `min-content` <= `max-content`.
 - As a max track sizing function, it behaves as `max-content`, except that it can be stretched at the end of the algorithm. So `min-content` <= `max-content` <= `auto`.

The example here uses `min-content, auto`, i.e. `mimax(min-content, min-content), minmax(auto, auto)`.
This is a somewhat contradictory input: the 1st track is requested to grow more as a minimum, but grow less a maximum. In CSS, minimums win, so that's why basically all of the contribution of the spanning item goes into the `min-content` track.

As Tab says, the effects vary a bit depending on the automatic minimum size. Removing the `overflow` distributes the size as a minimum contribution, so both rows get half of it since both are intrinsic. With `overflow`, the minimum contribution is 0, so the size is distributed as a min-content contribution, and it goes to the 1st track.

I don't think it's possible at this point to change the above. As fantasai says, to get the desired behavior, authors should use a different `grid-template-rows`, weakening the min track sizing function of the 1st track, and boosting the min track sizing function of the 2nd track.

```css
grid-template-rows: minmax(auto, min-content) minmax(min-content, auto);
```

That does the trick when the minimum contribution is 0, but otherwise the size will be distributed equally, just like with

```css
grid-template-rows: min-content minmax(min-content, auto);
grid-template-rows: minmax(auto, min-content) auto;
```

And it's at this point where I think we can improve the situation by doing what I said in https://github.com/w3c/csswg-drafts/issues/2873#issuecomment-1641000730

The growth limits have been set tight to the size of the non-spanning items, so if we prioritize respecting the limit of a track with a `min-content` max track sizing function over respecting the limit of a track with a `max-content` or `auto` max track sizing function, the extra contribution of the spanning item will go into the 2nd track.
